src/refract/ExpandVisitor.cc

```diff
@@ -0,0 +1,270 @@
+//
+//  refract/ExpandVisitor.cc
+//  librefract
+//
+//  Created by Jiri Kratochvil on 21/05/15.
+//  Copyright (c) 2015 Apiary Inc. All rights reserved.
+//
+
+#include "Element.h"
+#include "Visitors.h"
+#include "Registry.h"
+
+namespace refract
+{
+
+    namespace 
+    {
+
+        bool Expandable(const IElement& e) 
+        {
+            IsExpandableVisitor v;
+            e.content(v);
+            return v.get();
+        }
+
+        IElement* ExpandOrClone(const IElement* e, const Registry& registry) 
+        {
+            IElement* result = NULL;
+            if (!e) {
+                return result;
+            }
+
+            ExpandVisitor expander(registry);
+            e->content(expander);
+            result = expander.get();
+
+            if (!result) {
+                result = e->clone();
+            }
+
+            return result;
+        }
+
+        void CopyMetaId(IElement& dst, const IElement& src)
+        {
+            IElement::MemberElementCollection::const_iterator name = src.meta.find("id");
+            if (name != src.meta.end() && (*name)->value.second && !(*name)->value.second->empty()) {
+                dst.meta["id"] = (*name)->value.second->clone();
+            }
+        }
+
+        IElement* FindMemberByKey(const ObjectElement& e, const std::string& name)
+        {
+            for (ObjectElement::ValueType::const_iterator it = e.value.begin()
+                ; it != e.value.end()
+                ; ++it ) {
+
+                ComparableVisitor cmp(name, ComparableVisitor::key);
+                (*it)->content(cmp);
+
+                if (cmp.get()) { // key was recognized - it is save to cast to MemberElement
+                    MemberElement* m = static_cast<MemberElement*>(*it);
+                    return m->value.second;
+                }
+            }
+            return NULL;
+        }
+
+        typedef enum {
+            Inherited,
+            Referenced,
+        } NamedTypeExpansion;
+
+        template<typename T>
+        T* ExpandNamedType(const Registry& registry, const std::string& name, NamedTypeExpansion expansion)
+        {
+            std::string en = name;
+            T* e = NULL;
+            if (expansion == Inherited) {
+                e = new T;
+            }
+
+            // FIXME: add check against recursive inheritance
```

Comment:
I mean it should stay there:
- in future we will need to make it, so instead of searching where to make change we will know place.
- there can be someone who will use refract as separated library an he will not be protected about recursion.